Streaming PUT and GET requests

hcarty opened the pull request:
Fixes #4 

To minimize code changes and avoid losing existing integrity checks, potentially streaming uploads must include a pre-specified length and SHA256 hash.

I'm open to name changes/suggestions for the new functions.

hcarty committed:
Support streaming PUT bodies

hcarty committed:
Support streaming GET bodies

hcarty committed:
Update changelog

andersfugmann commented:
Thanks. Nice changes - I was actually expecting much more invasive changes, so I'm pleasantly surprised. 

I would prefer to create a module 'Streaming' where the data argument is of type 'Cohttp_async.Body', and length and an optional sha256 is given as arguments to the upload function(s), rather than as a separate data structure. If no sha256 is supplied for the data, an 'Unsigned payload option' should be used, and  the system should validate that the resulting md5 matches what has been sent (which in terms means that we should calculate the md5 on the fly). 

I'm eager to hear your comments on this. 

I do not mind doing the changes mentioned above, but I wont have time until sometime next week.

andersfugmann commented:
Just reading up on the Cohttp_lwt/async.Body.t type. The type for body should of course be stream or pipe and defined in compat.mli.

hcarty commented:
Moving these functions into a streaming module (or changing the names to `(put|get)_streaming`) sounds fine to me.  One benefit to having them alongside the existing `put` and `get` is that they're a little more discoverable.  I don't have strong feelings about that one way or the other though.

I agree that removing the `body` type and providing the fields as separate arguments, including making the sha256 optional, makes sense.

For the stream type, I think it would be ok to stick with the existing `Body.t` in compat.mli.  For both the Lwt and Async implementations, that type already includes the appropriate stream type.  It may be somewhat nonsensical in a localized sense to use the `string` or `string list` variants inherited from `Cohttp.Body.t`, but allowing the full type does make it easier to directly interact with other cohttp client and server code.

I'll try to at least get started on these changes today.

hcarty commented:
After thinking more about making the sha256 optional, I'm not as comfortable with that approach.  It makes it easy/easier to accidentally overwrite an existing value in a bucket with corrupt data.  It also requires reporting an error to the caller which can only happen when the sha256 is left out.  And it can leave "bad" data behind on S3.

What do you think of those concerns?

hcarty committed:
Make streaming body fields explicit arguments

hcarty committed:
Rename *_body functions to *_stream

hcarty commented:
I made some small changes.  The `body` type is no longer exposed - the fields are now split out as explicit arguments.  I also renamed `*_body` to `*_stream` to make their purpose more explicit.  I'd like your comments on this iteration and my concerns [above](https://github.com/andersfugmann/aws-s3/pull/5#issuecomment-400856424) before I make any more changes.  Thank you for the review and comments provided so far.

To your earlier/original comment - I'm quite happy with how small the changes required are too!  I wasn't expecting even basic support to be so straightforward.

hcarty committed:
Fix cohttp body type compatibility

hcarty committed:
Return content-length when calling get_stream

andersfugmann commented:
> For the stream type, I think it would be ok to stick with the existing Body.t in compat.mli. For both the Lwt and Async implementations, that type already includes the appropriate stream type. It may be somewhat nonsensical in a localized sense to use the string or string list variants inherited from Cohttp.Body.t, but allowing the full type does make it easier to directly interact with other cohttp client and server code.

Looking at the internals of the CoHttp.Body, I was thinking of making a type 
`Compat.Stream.t = string Async.Pipe.Reader.t` or
`Compat.Stream.t = string Lwt_stream.t`

That aws-s3 is using cohttp is an implementation detail that should be hidden from the users (Someday we might be using http_af).

As long as get and put can be 'chained'. 

> After thinking more about making the sha256 optional, I'm not as comfortable with that approach. It makes it easy/easier to accidentally overwrite an existing value in a bucket with corrupt data. It also requires reporting an error to the caller which can only happen when the sha256 is left out. And it can leave "bad" data behind on S3.

I agree that this is a concern, but I do see huge advantages in allowing true streaming. Given a stream type `t`, maybe the body could have the type: 
`type body = Signed of sha256 * Stream.t | Unsigned of Stream.t.`. This way its very explicit that its an unsigned payload. 

It should be up to the user to decide how to handle broken uploads (Uploads with the wrong sha), and we should state the pitfalls in the documentation. Come think of it, using multipart uploads can be made safe wrt. streaming I think, as we can validate the md5(s) before ending the multipart upload.  

(And we can create helpers later to convert an unsigned stream to a signed one).

Another thing we should consider: Given that data is streamed, and we don't want to hold on to the data, we should support _100-continue_, where the body is only sent after getting initial ack from s3. This allows the code to handle redirects and other errors before sending the body. But that can come later.

andersfugmann commented:
Also - I just noticed that aws supports 'chunked uploads', which seems to be exactly what we are doing here. See https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-streaming.html

We might consider that, as we can then avoid the problem of bad uploads (broken files being left behind. 

This is orthogonal to using 100-continue.

hcarty commented:
I'm going to close this in favor of #8 (which I haven't had a chance to review yet!) due to the issues raised about cohttp's streaming support.  I'll leave the branch in place for now in case it's useful at some point.

Thank you for the review and for continuing work on this!
